### ISSUE TITLE
Single PointLight2D causes extreme lag on Android

### ISSUE THREAD

**Issue by udit** [godotengine/godot#81152]
### Godot version

v4.1.1.stable.official [bd6af8e0e]

### System information

Android 13 - Vulkan (Mobile), Device - [Redmi Note 7 Pro](https://www.gsmarena.com/xiaomi_redmi_note_7_pro-9521.php), CPU - Qualcomm Snapdragon 675 Octa-core upto 2.0GHz, GPU - Adreno 612 GPU upto 845MHz, Display - 2340x1080 FHD+ 409 PPI

### Issue description

I made a test scene with 3 nodes - Fullscreen `ColorRect`, `Sprite2D`, `PointLight2D`. Whenever the light is in the active area, FPS reduces to 42 from 60. Is this an expected behaviour with Light2D? Seems like either I am doing something totally wrong or the light nodes are not at all usable for mobile games.

![Screenshot_20230830-123903](https://github.com/godotengine/godot/assets/6249061/d05091f3-6cdc-42a6-90a9-d96dd02809d1)

60 FPS with light off:
![Screenshot_20230830-124715](https://github.com/godotengine/godot/assets/6249061/eed5e279-10dd-449b-a421-b45bd410242c)

### Steps to reproduce

Attaching the test project below. Just one click run on an android device.

### Minimal reproduction project

[Light2DTest.zip](https://github.com/godotengine/godot/files/12472776/Light2DTest.zip)

**Comment by Calinou** [comment]
The Vulkan Mobile rendering backend is high-end oriented. Try using the Compatibility rendering method instead.

**Comment by Calinou** [comment]
> Using compatibility rendering method didn't have any effect on frame rate (I assume it's the 'Compatibility' option in the top right corner.).

[Unfortunately, no.](https://github.com/godotengine/godot/pull/72461) You need to open the Project Settings and change **Rendering Method.mobile** to `gl_compatibility`.

**Comment by Zireael07** [comment]
The biggest thing I can think of is inability to use instance shader parameters (they can come useful in 2d not just 3d).

As for your question, IIRC there's an open proposal for something like that.

**Comment by Calinou** [comment]
> Are there any drawbacks of exporting 2d project with gl_compatibility? I guess this is not something that can be changed after exporting project? Like encapsulating this option into something like "high graphics" and "low graphics" option to end users?

For mobile, we recommend Compatibility as it's the only rendering method that works reliably on older/low-end devices. The Mobile rendering method is high-end-oriented.

It *is* possible to provide a run-time selection that is effective after a restart (by writing `user://override.cfg` with the appropriate setting using the ConfigFile class), but it's not trivial to support. You will effectively have two rendering codepaths to cater for in your project.

**Comment by clayjohn** [comment]
I did a bit of testing related to this on another PR and I have a diff that helps. On my device (Pixel 4) it brings the performance back up to 60 FPS. I modified the MRP a bit to make it stress the GPU more by adding 16 lights total

```
diff --git a/servers/rendering/renderer_rd/shaders/canvas.glsl b/servers/rendering/renderer_rd/shaders/canvas.glsl
index 235c772e2d..dbff09c301 100644
--- a/servers/rendering/renderer_rd/shaders/canvas.glsl
+++ b/servers/rendering/renderer_rd/shaders/canvas.glsl
@@ -665,6 +665,12 @@ void main() {
 
                vec2 tex_uv = (vec4(vertex, 0.0, 1.0) * mat4(light_array.data[light_base].texture_matrix[0], light_array.data[light_base].texture_matrix[1], vec4(0.0, 0.0, 1.0, 0.0), vec4(0.0, 0.0, 0.0, 1.0))).xy; //multiply inverse given its transposed. Optimizer removes useless operations.
                vec2 tex_uv_atlas = tex_uv * light_array.data[light_base].atlas_rect.zw + light_array.data[light_base].atlas_rect.xy;
+
+               if (any(lessThan(tex_uv, vec2(0.0, 0.0))) || any(greaterThanEqual(tex_uv, vec2(1.0, 1.0)))) {
+                       //if outside the light texture, light color is zero
+                       continue;
+               }
+
                vec4 light_color = textureLod(sampler2D(atlas_texture, texture_sampler), tex_uv_atlas, 0.0);
                vec4 light_base_color = light_array.data[light_base].color;
 
@@ -689,10 +695,6 @@ void main() {
                        light_color.rgb *= base_color.rgb;
                }
 #endif
-               if (any(lessThan(tex_uv, vec2(0.0, 0.0))) || any(greaterThanEqual(tex_uv, vec2(1.0, 1.0)))) {
-                       //if outside the light texture, light color is zero
-                       light_color.a = 0.0;
-               }
 
                if (bool(light_array.data[light_base].flags & LIGHT_FLAGS_HAS_SHADOW)) {
                        vec2 shadow_pos = (vec4(shadow_vertex, 0.0, 1.0) * mat4(light_array.data[light_base].shadow_matrix[0], light_array.data[light_base].shadow_matrix[1], vec4(0.0, 0.0, 1.0, 0.0), vec4(0.0, 0.0, 0.0, 1.0))).xy; //multiply inverse given its transposed. Optimizer removes useless operations.
```

_Before:_
![42fps](https://github.com/godotengine/godot/assets/16521339/28f2e853-144a-4ea5-9f75-f99fe742bf88)

_After:_

![60fps](https://github.com/godotengine/godot/assets/16521339/925431e0-b64d-4e08-8444-0e441333c32b)

**Comment by akien-mga** [comment]
A potential fix has been merged for this in #90920, please test 4.3-dev6 when it's released in a few days if you can, and confirm whether this solves the lag issue.
